[not verified] Refresh the autoloader after a plugin update.

enejb opened the pull request:
Currently when we update Jetpack and we move files around we end up throwing an PHP fatal because the classes are being included in a way that doesn't match the file system any more. 

This PR tires to fix this by hooking into the plugin update mechanism and updating mapping of the classes user by the autoloader.

One thing that currently is not considered the the way that the individual files are loaded twice. This is not idea because it means that in some cases we are loading the add action multiple times. 

I think a fix for this should be to move away from loading of any individual files. 

#### Changes proposed in this Pull Request:
* Hook into the plugin update and update the autoloader class mapping. 
* Update the unique function to an anonymous function so that it can be declared and executed multiple times. When we include the autoloader multiple times. 

#### Testing instructions:
* Do the tests pass? 
* I need help testing this. I am not sure what is a good way to do this. 

#### Proposed changelog entry for your changes:
* Improve the plugin update so that we end up with less fatal errors.

enejb committed:
[not verified] Refresh the autoloader after a plugin update.

enejb committed:
Remove the error log

jeherve commented on jetpack.php:
```diff
@@ -93,6 +93,21 @@ function jetpack_admin_unsupported_wp_notice() { ?>
+		return $worked;
+	}
+
 	require $jetpack_autoloader;
```
Since we may include it above, should we change this to `require_once`?

kbrown9 replied:
I think `require` is ok here. This file, `jetpack.php`, should only be loaded once, so this `require` is only called once. The `include` statement in line 112 is in the `jetpack_refresh_autoloader_post_install()` function, which is hooked to `upgrader_post_install`. It will be called after the `require`.

jeherve commented on jetpack.php:
```diff
@@ -93,6 +93,21 @@ function jetpack_admin_unsupported_wp_notice() { ?>
 if ( is_readable( $jetpack_autoloader ) ) {
+	add_filter( 'upgrader_post_install', 'jetpack_refresh_autoloader_post_install', 1, 2 );
+	function jetpack_refresh_autoloader_post_install( $worked, $hook_extras ) {
+		error_log( print_r( $hook_extras, 1 ) );
```
We can probably remove this now.

kbrown9 replied:
Fixed in c137956.

jeherve reviewed (commented):
I am not quite sure how to test this either, but left minor comments below for now.

zinigor commented:
Looks like this PR still needs some work, let's move it to 8.2 because 8.1 will not rename/move/delete files as much as 8.0.

kbrown9 committed:
Autoloader: code cleanup

- Change test class name to 'WP_Test_Jetpack_Include_Autoloader'.
- Fix PHPCS warnings.
- Remove an error_log().

kbrown9 committed:
Autoloader: fix a typo in a comment

'Shifter' should be 'shifted'.

kbrown9 commented:
I tested this PR, and the PHP fatal error was still generated when updating from version 7.9.1. I started with 7.9.1 because files were moved between versions 7.9.1 and 8.0.

The test steps were:

1. Install Jetpack version 7.9.1.
2. Connect to Jetpack.
3. Update to the [8.2-alpha version](https://plugins.trac.wordpress.org/browser/jetpack/tags/8.2-alpha) (which includes the changes in this PR ) using wp-cli:
`wp plugin update jetpack --version=8.2-alpha`

This fatal error was generated:

> Fatal error: Uncaught Error: Class 'Automattic\Jetpack\Roles' not found in …/jetpack/vendor/automattic/jetpack-sync/src/Listener.php:270
> Stack trace:
> \# 0 …jetpack/vendor/automattic/jetpack-sync/src/Listener.php(239): Automattic\Jetpack\Sync\Listener->get_actor('jetpack_plugin_...', Array)
> \# 1 …/jetpack/vendor/automattic/jetpack-sync/src/Listener.php(114): Automattic\Jetpack\Sync\Listener->enqueue_action('jetpack_plugin_...', Array, Object(Automattic\Jetpack\Sync\Queue))
> \# 2 …/wp-includes/class-wp-hook.php(288): Automattic\Jetpack\Sync\Listener->action_handler(Array)
> \# 3 …/wp-includes/class-wp-hook.php(312): WP_Hook->apply_filters('', Array)
> \# 4 …/wp-includes/plugin.php(478): WP_Hook->do_action in …/jetpack/vendor/automattic/jetpack-sync/src/Listener.php on line 270
> Error: There has been a critical error on your website.Learn more about debugging in WordPress. There has been a critical error on your website.

kbrown9 commented on jetpack.php:
```diff
@@ -93,6 +93,27 @@ function jetpack_admin_unsupported_wp_notice() { ?>
+	function jetpack_refresh_autoloader_post_install( $worked, $hook_extras ) {
+		if (
+			! isset( $hook_extras['plugin'] )
+			|| JETPACK__PLUGIN_FILE !== $hook_extras['plugin']
```
Based on the description of the `$plugin` parameter [here](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/class-plugin-upgrader.php#L138) and some testing, I don't think `JETPACK__PLUGIN_FILE` will work here. `JETPACK__PLUGIN_FILE` is the full path, but it looks like `$hook_extras['plugin']` will contain the plugin file path relative to the plugins directory.

enejb commented:
I am closing this since there will be a new Jetpack Autoupdater and this PR got stale.
